config/config.go

```diff
@@ -294,6 +306,83 @@ func (m *ManualWarpMessage) Validate() error {
 	return nil
 }
 
+// If the numerator in the Warp config is 0, use the default value
+func calculateQuorumNumerator(cfgNumerator uint64) uint64 {
+	if cfgNumerator == 0 {
+		return warp.WarpDefaultQuorumNumerator
+	}
+	return cfgNumerator
+}
+
+// Helper to retrieve the Warp Quorum from the chain config.
+// Differentiates between subnet-evm and coreth RPC internally
+func getWarpQuorum(
+	subnetID ids.ID,
+	blockchainID ids.ID,
+	client ethclient.Client,
+) (WarpQuorum, error) {
+	if subnetID == constants.PrimaryNetworkID {
+		return WarpQuorum{
+			QuorumNumerator:   warp.WarpDefaultQuorumNumerator,
+			QuorumDenominator: warp.WarpQuorumDenominator,
+		}, nil
+	}
+
+	// Fetch the subnet's chain config
+	chainConfig, err := client.ChainConfig(context.Background())
+	if err != nil {
+		return WarpQuorum{}, fmt.Errorf("failed to fetch chain config for blockchain %s: %v", blockchainID, err)
+	}
+
+	// First, check the list of precompile upgrades to get the most up to date Warp config
+	// We only need to consider the most recent Warp config, since the QuorumNumerator is used
+	// at signature verification time on the receiving chain, regardless of the Warp config at the
+	// time of the message's creation
+	var warpConfig *warp.Config
+	for _, precompile := range chainConfig.UpgradeConfig.PrecompileUpgrades {
+		cfg, ok := precompile.Config.(*warp.Config)
+		if !ok {
+			continue
+		}
+		if warpConfig == nil {
+			warpConfig = cfg
+			continue
+		}
+		if *cfg.Timestamp() > *warpConfig.Timestamp() {
+			warpConfig = cfg
+		}
+	}
+	if warpConfig != nil {
+		return WarpQuorum{
+			QuorumNumerator:   calculateQuorumNumerator(warpConfig.QuorumNumerator),
+			QuorumDenominator: warp.WarpQuorumDenominator,
+		}, nil
+	}
+
+	// If we didn't find the Warp config in the upgrade precompile list, check the genesis config
+	warpConfig, ok := chainConfig.GenesisPrecompiles[warpConfigKey].(*warp.Config)
+	if ok {
+		return WarpQuorum{
+			QuorumNumerator:   calculateQuorumNumerator(warpConfig.QuorumNumerator),
+			QuorumDenominator: warp.WarpQuorumDenominator,
+		}, nil
+	}
+	return WarpQuorum{}, fmt.Errorf("failed to find warp config for blockchain %s", blockchainID)
+}
+
+func (c *Config) InitializeWarpQuorums() error {
+	// Fetch the Warp quorum values for each destination subnet.
+	for i, destinationSubnet := range c.DestinationSubnets {
+		err := destinationSubnet.initializeWarpQuorum()
+		if err != nil {
+			return fmt.Errorf("failed to initialize Warp quorum for destination subnet %s: %v", destinationSubnet.SubnetID, err)
+		}
+		c.DestinationSubnets[i] = destinationSubnet
```

Comment:
We do have to rewrite to `c` since `destinationSubnet` is a copy of the slice element, and not a reference to the element itself. AFAIK Go has no way to iterate over a slice by reference (compare to `auto& [x, y]` in C++ for example).

That said, I did not realize that Go's JSON package could unmarshal into pointers. I've changed `Config` to be composed of  slices of pointers so that we can iterate over the slice by reference and skip having to write back to the `Config` object.